src/vtt.c

```diff
@@ -0,0 +1,240 @@
+            trimmed_length = utf8_trimmed_length(data, line_length);
+            // Skip empty lines
+        } while (0 < line_length && 0 == trimmed_length);
+
+        // linelength cant be zero before EOF
+        if (0 == line_length) {
+            break;
+        }
+
+        if (strnstr(data, "-->", line_length) != NULL) {
+            fprintf(stderr, "Parse timestamps\n");
+            parse_timestamps(data, &str_pts, &end_pts, &cue_settings);
+            if (str_pts == -1) {
+                // Failed to parse timestamps
+                printf("Bad timestamp\n");
+                return NULL;
+            }
+        } else {
+            // Expected timestamp
+            printf("Expected timestamp\n");
+            return NULL;
+        }
+        
+        data += line_length;
+        size -= line_length;
+
+        // Caption text starts here
+        const utf8_char_t* text = data;
+        size_t text_size = 0;
+
+        line_length = 0;
+
+        do {
+            text_size += line_length;
+            line_length = utf8_line_length(data);
+            trimmed_length = utf8_trimmed_length(data, line_length);
+            // printf ("cap (%d): '%.*s'\n", line_length, (int) trimmed_length, data);
+            data += line_length;
+            size -= line_length;
+        } while (trimmed_length);
+
+        // should we trim here?
+        vtt_t* vtt = vtt_new(text, text_size, str_pts, prev, &head);
+        prev = vtt;
+        vtt->duration = end_pts - str_pts;
+        vtt->cue_settings = cue_settings;
+    }
+
+    return head;
+}
+
+int vtt_to_caption_frame(vtt_t* vtt, caption_frame_t* frame)
+{
+    const char* data = vtt_data(vtt);
+    return caption_frame_from_text(frame, data);
+}
+
+vtt_t* vtt_from_caption_frame(caption_frame_t* frame, vtt_t* prev, vtt_t** head)
+{
+    // CRLF per row, plus an extra at the end
+    vtt_t* vtt = vtt_new(NULL, 2 + CAPTION_FRAME_TEXT_BYTES, frame->timestamp, prev, head);
+    utf8_char_t* data = vtt_data(vtt);
+
+    caption_frame_to_text(frame, data);
+    // vtt requires an extra new line
+    strcat((char*)data, "\r\n");
+
+    return vtt;
+}
+
+static inline void _crack_time(double tt, int* hh, int* mm, int* ss, int* ms)
+{
+    (*ms) = (int)((int64_t)(tt * 1000) % 1000);
+    (*ss) = (int)((int64_t)(tt) % 60);
+    (*mm) = (int)((int64_t)(tt / (60)) % 60);
+    (*hh) = (int)((int64_t)(tt / (60 * 60)));
+}
+
+static void _dump(vtt_t* head)
+{
+    int i;
+    vtt_t* vtt;
+
+    printf("WEBVTT\r\n");
+
+    for (vtt = head, i = 1; vtt; vtt = vtt_next(vtt), ++i) {
+        int hh1, hh2, mm1, mm2, ss1, ss2, ms1, ms2;
+        _crack_time(vtt->timestamp, &hh1, &mm1, &ss1, &ms1);
+        _crack_time(vtt->timestamp + vtt->duration, &hh2, &mm2, &ss2, &ms2);
+
+        printf("%d:%02d:%02d.%03d --> %02d:%02d:%02d.%03d\r\n%s\r\n",
+            hh1, mm1, ss1, ms1, hh2, mm2, ss2, ms2, vtt_data(vtt));
+    }
+}
+
+void vtt_dump(vtt_t* head) { _dump(head); }
```

Comment:
Lets share as much code as we can with srt.

Reply:
👍